### PR TITLE
Move (old) checkpoint creation logic to `Cardano.Wallet.Checkpoints`

### PR DESCRIPTION
**Pull request by HeinrichApfelmus**
### Issue number

ADP-1043

### Overview

This pull request is a pure refactoring.

We move the values and types
* `SparseCheckpointsConfig`
* `defaultSparseCheckpointsConfig`
* `sparseCheckpoints`
* `gapSize`

out of the `Cardano.Wallet.DB` module and into the `Cardano.Wallet.Checkpoints` module.

### Comments

This is still the old checkpointing logic! We only move it around here — it will be replaced later. Divide and conquer. ⚔️

### CODE REVIEW

**Review by paolino — APPROVED** [review]
LGTM

File: lib/core/src/Cardano/Wallet/Checkpoints.hs [hunk 1]
```diff
@@ -132,3 +145,120 @@ instance Buildable (DeltaCheckpoints a) where
     build (PutCheckpoint slot _) = "PutCheckpoint " <> build slot
     build (RollbackTo slot) = "RollbackTo " <> build slot
     build (RestrictTo slots) = "RestrictTo " <> listF slots
+
+{-------------------------------------------------------------------------------
+    Checkpoint hygiene
+-------------------------------------------------------------------------------}
+-- | Storing EVERY checkpoints in the database is quite expensive and useless.
```
Comment (paweljakubas) [inline lib/core/src/Cardano/Wallet/Checkpoints.hs:152]: I am not 100% sure it is useless, but I agree it is expensive - heheh

File: lib/core/src/Cardano/Wallet/Checkpoints.hs [hunk 2]
```diff
@@ -132,3 +145,120 @@ instance Buildable (DeltaCheckpoints a) where
...
+-- We make the following assumptions:
+--
+-- - We can't rollback for more than `k=epochStability` blocks in the past
+-- - It is pretty fast to re-sync a few hundred blocks
+-- - Small rollbacks may occur more often than long one
```
Comment (paweljakubas) [inline lib/core/src/Cardano/Wallet/Checkpoints.hs]:
```suggestion
-- - Small rollbacks may occur more often than deep ones
```

File: lib/core/src/Cardano/Wallet/Checkpoints.hs [hunk 3]
```diff
@@ -132,3 +145,120 @@ instance Buildable (DeltaCheckpoints a) where
...
+-- │cp000 │cp100 │cp101 │..    ..│cp110 │
+-- └───┴───┴───┴─  ──┴───┘
+--  Want in DB:
+-- ┌───┬───┬───┬─  ──┬───┐
+-- │cp000 │cp100 │cp101 │..    ..│cp111 │
+-- └───┴───┴───┴─  ──┴───┘
+--  @
+--
+-- NOTE: There might be cases where the chain following "fails" (because, for
+-- example, the node has switch to a different chain, different by more than k),
```
Comment (paweljakubas) [inline lib/core/src/Cardano/Wallet/Checkpoints.hs]:
```suggestion
-- example, the node has switched to a different chain, different by more than k),
```

File: lib/core/src/Cardano/Wallet/Checkpoints.hs [hunk 4]
```diff
@@ -132,3 +145,120 @@ instance Buildable (DeltaCheckpoints a) where
...
+            then [0..h]
+            else [h-e,h-e+1..h]
+    in
+        L.sort (L.nub $ initial : (longTerm ++ shortTerm))
+
+-- | Captures the configuration for the `sparseCheckpoints` function.
+--
+-- NOTE: large values of 'edgeSize' aren't recommended as they would mean
+-- storing many unnecessary checkpoints. In Ouroboros Praos, there's a
+-- reasonable probability for small forks of a few blocks so it makes sense to
```
Comment (paweljakubas) [inline lib/core/src/Cardano/Wallet/Checkpoints.hs]:
```suggestion
-- reasonable probability for small forks each a few blocks deep so it makes sense to
```

**Review by paweljakubas — APPROVED** [review]
looks like a reasonable code reorganization

File: lib/core/src/Cardano/Wallet/Checkpoints.hs (hunk 1 above)
Comment (paweljakubas) [inline lib/core/src/Cardano/Wallet/Checkpoints.hs:152]: well, I agree in 100% it is costly, but not 100% that it is useless - what about blockchain having 3 checkpoints depth? :trollface:

Comment (HeinrichApfelmus) [reply lib/core/src/Cardano/Wallet/Checkpoints.hs:152]: I will rephrase this comment from the past in a future pull request. 🤓

**Comment by HeinrichApfelmus** [comment]
bors merge

**Pull request merged**